electrum/bip39_recovery.py

```diff
@@ -0,0 +1,51 @@
+# Copyright (C) 2020 The Electrum developers
+# Distributed under the MIT software license, see the accompanying
+# file LICENCE or http://www.opensource.org/licenses/mit-license.php
+
+from . import bitcoin
+from .constants import BIP39_WALLET_FORMATS
+from .keystore import bip39_to_seed
+from .bip32 import BIP32_PRIME, BIP32Node
+from .bip32 import convert_bip32_path_to_list_of_uint32 as bip32_str_to_ints
+from .bip32 import convert_bip32_intpath_to_strpath as bip32_ints_to_str
+
+async def account_discovery(network, mnemonic, passphrase=""):
+    seed = bip39_to_seed(mnemonic, passphrase)
+    root_node = BIP32Node.from_rootseed(seed, xtype="standard")
+    active_accounts = []
+    for wallet_format in BIP39_WALLET_FORMATS:
+        account_path = bip32_str_to_ints(wallet_format["derivation_path"])
+        while True:
+            account_node = root_node.subkey_at_private_derivation(account_path)
+            has_history = await account_has_history(network, account_node, wallet_format["script_type"]);
+            if has_history:
+                account = format_account(wallet_format, account_path)
+                active_accounts.append(account)
+            if not has_history or not wallet_format["iterate_accounts"]:
+                break
+            account_path[-1] = account_path[-1] + 1
+    return active_accounts
+
+async def account_has_history(network, account_node, script_type):
+    gap_limit = 20
+    for address_index in range(gap_limit):
+        address_node = account_node.subkey_at_public_derivation("0/" + str(address_index))
+        pubkey = address_node.eckey.get_public_key_hex()
+        address = bitcoin.pubkey_to_address(script_type, pubkey)
+        script = bitcoin.address_to_script(address)
+        scripthash = bitcoin.script_to_scripthash(script)
+        history = await network.get_history_for_scripthash(scripthash)
+        if len(history) > 0:
+            return True
```

Comment:
This `for` [...] `await task` construct here will mean that we send a request to the server, wait for it to answer, and then send the next request. This is why it is taking as long as you said (several seconds or similar).
That is what the `await` does, it actually stops the coroutine, and asks to only be rescheduled after that task has returned (mid-loop).

Instead, I suggest you use a TaskGroup. A TaskGroup lets you create a context (a `with` block) which will only exit once all the tasks inside returned; or if any task raises, it will cancel the remaining tasks and just re-raise the exception (so that it can propagate out).
That way, you can send all the requests "concurrently", i.e. send them all, and handle the responses as they arrive.
See these examples:
https://github.com/spesmilo/electrum/blob/fcbc1c9a459d79d54d6c102912f56b922d58fb19/electrum/interface.py#L567
https://github.com/spesmilo/electrum/blob/fcbc1c9a459d79d54d6c102912f56b922d58fb19/electrum/interface.py#L594
https://github.com/spesmilo/electrum/blob/fcbc1c9a459d79d54d6c102912f56b922d58fb19/electrum/tests/test_lnpeer.py#L504

Note that atm these nested loops you have will send all N*20 requests in series, so we will wait for over 100 network RTTs. By putting a TaskGroup here, you can reduce that to around N RTTs (N being the number of derivation path prefixes).
As further optimisation, you could use a TaskGroup at the outer loop too, so that it becomes just a single RTT (disregarding the account stuff for the moment).

Reply:
Ahh, thanks!

I actually attempted to get all requests going concurrently by first saving executing the async functoins and saving the return values in a variable and then awaiting them later.

Something like:

```python
requests = []
for url in urls:
  requests.append(fetch(url))
results = []
for request in requests:
  results.append(await request)
```

This works in JavaScript since the async function is triggered by the function execution, `await` is just used to wait for the response.

Seems like in Python the async function execution just returns some "thing" that represents an async function call and doesn't actually execute anything until it is awaited, is that correct?

Since the above code still seemed to execute the requests serially.

e.g, equiv JavaScript code is:

```js
equests = []
for (url of urls) {
    requests.push(fetch(url))
}
results = []
for (request of requests) {
    results.push(await request)
}
```

All `fetch()` requests actually happen concurrently here and once the slowest request has completed you and up with `results` being an array of all the resolved values.

I'll have a play with `TaskGroup` and try and get it working concurrently.

Also, is there not a risk we hit ElectrumX rate limits if we do all requests concurrently?

Reply:
I implemented concurrency, both at the address lookup level and the derivation scheme level.

The entire discovery process now takes ~2s for my test seed. 🎉